### PR TITLE
Dont report the same backtrace from multiple members in the source chain

### PR DESCRIPTION
**Pull request by yaahc**

### CODE REVIEW

File: src/wrapper.rs [hunk 1]
```diff
@@ -4,6 +4,8 @@ use core::fmt::{self, Debug, Display};
 #[repr(transparent)]
 pub struct MessageError<M>(pub M);
 
+// TODO add backtrace member to message error
```
Comment (yaahc) [inline src/wrapper.rs:7]: If we dont capture a backtrace in the root error then an iteration over the inner errors will not yield _any_ backtraces on any of the errors in the chain, 

with the changes above when serializing the inner errors we get:

```
{
  "msg": "outermost error",
  "backtrace": "disabled backtrace",
  "source": {
    "msg": "second error",
    "backtrace": "disabled backtrace",
    "source": {
      "msg": "root cause",
      "backtrace": null,
      "source": null
    }
  }
}
```

Whereas with the changes and without the backtrace in Message error we get:

```
{
  "msg": "outermost error",
  "backtrace": null,
  "source": {
    "msg": "second error",
    "backtrace": null,
    "source": {
      "msg": "root cause",
      "backtrace": null,
      "source": null
    }
  }
}
```

**Comment by yaahc** [comment]
> In general I don't have a problem with multiple errors in a chain giving the same backtrace. This would happen e.g. if downstream defines an error type like:

Yea, there's no way to prevent users from doing this but I still feel like it's an anti pattern in the same way that including your source's error message in your `Display` output is an anti pattern, you lose the context on where that Backtrace comes from in a chain of errors.

I'm not sure it matters with backtraces at all. But I can imagine a situation where someone is sending errors across threads and wants to support capturing and displaying multiple Backtrace, from the different threads it propagates through, in the final error report. I feel like the current behaviour makes `.context` incompatible with usecases like that.

> For an error serializer, maybe you could use https://doc.rust-lang.org/std/ptr/fn.eq.html on consecutive backtraces to decide whether to keep only one?

This is good advice, ty.

**Comment by dtolnay** [comment]
> you lose the context on where that Backtrace comes from in a chain of errors.

I don't think you really lose this information, since ptr::eq can cheaply recover which is the innermost error associated with the backtrace. An inner error couldn't return a reference to a backtrace belonging to an outer error.

> I can imagine a situation where someone is sending errors across threads and wants to support capturing and displaying multiple Backtrace, from the different threads it propagates through, in the final error report. I feel like the current behavior makes `.context` incompatible with usecases like that.

Is this saying you'd like every `context` call to capture a new backtrace? I don't think that would be the right default behavior for `context`. But a codebase that needs this could do it with their own context extension trait instead of anyhow::Context.

**Comment by yaahc** [comment]
> Is this saying you'd like every context call to capture a new backtrace?

I was thinking Option::take the backtrace from the inner Error

**Comment by dtolnay** [comment]
> I can imagine a situation where someone is sending errors across threads and wants to support capturing and displaying multiple Backtrace, from the different threads it propagates through, in the final error report.

> I was thinking Option::take the backtrace from the inner anyhow::Error

Then there's still only one backtrace for the final report right, not one per thread it propagates across? Capturing an additional backtrace would still require an alternative context extension trait like today.

**Comment by yaahc** [comment]
> Then there's still only one backtrace for the final report right, not one per thread it propagates across? Capturing an additional backtrace would still require an alternative context extension trait like today.

This is a good point, anyhow isn't gonna support printing the multiple backtraces or capturing them, so you'd have to do all the capturing of extra backtraces in custom Error types and then you wouldn't be using `anyhow` as a reporter in that case.

Thank you for entertaining my errant error handling thoughts, I'll probably add the bit about ptr eq to my rfc for member access, because I was worried about designs that would encourage forwarding member access to the source and had already mentioned it.

In a vaguely related comment

```rust
macro_rules! backtrace_if_absent {
    ($err:expr) => {
        match $err.backtrace() {
            Some(_) => None,
            None => Some(Backtrace::capture()),
        }
    };
}
```

This will only not capture a backtrace if the outermost error in a chain has a backtrace right? How would you feel about me making a PR to do `err.chain().rev().find(|e| e.backtrace())` so that it finds any backtrace, and does so from the root cause first so it gets hopefully the deepest backtrace in the chain on the off chance someone is capturing multiple backtraces.

**Comment by dtolnay** [comment]
Feel free to send a PR. I may need to consider a benchmark before accepting because that is quadratically expensive in the length of the chain if each backtrace method forwards to the next error's backtrace method.

**Pull request closed without merging**